File: src/main/java/com/github/nsc/de/shake/lexer/characterinput/charactersource/CharacterSource.java
package com.github.nsc.de.shake.lexer.characterinput.charactersource;

import java.io.*;

public interface CharacterSource {

    char[] get(int start, int end);
    char[] getAll();
    int getLength();
    String getLocation();

    static CharacterSource from(String s, String source) {
        return from(s.toCharArray(), source);
    }

    static CharacterSource from(char[] chars, String source) {
        return new CharArraySource(chars, source);
    }

    static CharacterSource from(File f, String source) throws IOException {
        BufferedReader reader = new BufferedReader(new FileReader(f));
        char[] chars = new char[(int) f.length()]; // FIXME files that are longer than 2^31 (integer limit)
        for(int i = 0; i < chars.length; i++) chars[i] = (char) reader.read();
        return from(chars, source);
    }

    static CharacterSource from(InputStream s, String source) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(s));
        char[] chars = new char[(int) s.available()]; // FIXME files that are longer than 2^31 (integer limit)
        for(int i = 0; i < chars.length; i++) chars[i] = (char) reader.read();
        return from(chars, source);
    }

}

File: util/src/main/kotlin/com/github/nsc/de/shake/util/characterinput/charactersource/CharacterSource.kt
package com.github.nsc.de.shake.util.characterinput.charactersource

import java.io.*

interface CharacterSource {

    operator fun get(start: Int, end: Int): CharArray
    val all: CharArray
    val length: Int
    val location: String

    companion object {

        @JvmStatic
        fun from(s: String, source: String): CharacterSource {
            return from(s.toCharArray(), source)
        }

        @JvmStatic
        fun from(chars: CharArray, source: String): CharacterSource {
            return CharArraySource(chars, source)
        }

        @JvmStatic
        @Throws(IOException::class)
        fun from(f: File, source: String): CharacterSource {
            val reader = BufferedReader(FileReader(f))
            val chars = CharArray(f.length().toInt()) // FIXME files that are longer than 2^31 (integer limit)
            for (i in chars.indices) chars[i] = reader.read().toChar()
            return from(chars, source)
        }

        @JvmStatic
        @Throws(IOException::class)
        fun from(s: InputStream, source: String): CharacterSource {
            val reader = BufferedReader(InputStreamReader(s))
            val chars = CharArray(s.available()) // FIXME files that are longer than 2^31 (integer limit)
            for (i in chars.indices) chars[i] = reader.read().toChar()
            return from(chars, source)
        }

    }
}